Source/SwiftLintFramework/Rules/Metrics/LargeTupleRule.swift

```diff
@@ -24,152 +14,29 @@ public struct LargeTupleRule: ASTRule, ConfigurationProviderRule {
-            let nameLength = dictionary.nameLength,
-            let length = dictionary.length,
-            let offset = dictionary.offset else {
                 return nil
-        }
-
-        let start = nameOffset + nameLength
-        let end = dictionary.bodyOffset ?? length + offset
-
-        guard end - start > 0 else {
-            return nil
-        }
-
-        return ByteRange(location: start, length: end - start)
-    }
-
-    private func parenthesesRanges(in text: String) throws -> [(NSRange, RangeKind)] {
-        var stack = [(Int, String)]()
-        var balanced = true
-        var ranges = [(NSRange, RangeKind)]()
-
-        let nsText = text.bridge()
-        let parenthesesAndAngleBrackets = CharacterSet(charactersIn: "()<>")
-        var index = 0
-        let length = nsText.length
-
-        while balanced {
-            let searchRange = NSRange(location: index, length: length - index)
-            let range = nsText.rangeOfCharacter(from: parenthesesAndAngleBrackets,
-                                                options: [.literal], range: searchRange)
-            if range.location == NSNotFound {
-                break
             }
-
-            index = NSMaxRange(range)
-            let symbol = nsText.substring(with: range)
-
-            // skip return arrows
-            if symbol == ">",
-                case let arrowRange = nsText.range(of: "->", options: [.literal], range: searchRange),
-                arrowRange.intersects(range) {
-                continue
-            }
-
-            if symbol == "(" || symbol == "<" {
-                stack.append((range.location, symbol))
-            } else if let (startIdx, previousSymbol) = stack.popLast(),
-                isBalanced(currentSymbol: symbol, previousSymbol: previousSymbol) {
-                let range = NSRange(location: startIdx, length: range.location - startIdx + 1)
-                let kind: RangeKind = symbol == ")" ? .tuple : .generic
-                ranges.append((range, kind))
-            } else {
-                balanced = false
-            }
-        }
-
-        guard balanced && stack.isEmpty else {
-            throw LargeTupleRuleError.unbalencedParentheses
-        }
-
-        return ranges
-    }
-
-    private func isBalanced(currentSymbol: String, previousSymbol: String) -> Bool {
-        return (currentSymbol == ")" && previousSymbol == "(") ||
-            (currentSymbol == ">" && previousSymbol == "<")
     }
+}
 
-    private func containsReturnArrow(in text: String, range: NSRange) -> Bool {
-        let arrowRegex = SwiftVersion.current >= .fiveDotFive
-                        ? regex("\\A(?:\\s*async)?(?:\\s*throws)?\\s*->")
-                        : regex("\\A(?:\\s*throws)?\\s*->")
-        let start = NSMaxRange(range)
-        let restOfStringRange = NSRange(location: start, length: text.bridge().length - start)
+private final class LargeTupleRuleVisitor: SyntaxVisitor {
```

Comment:
Another great simplification thanks to SwiftSyntax!

By the way, I'd love to make improvements to SwiftSyntaxRule and the related SwiftSyntax convenience protocols so that this case could be covered. Let me know if you think of a way to do that!

Reply:
Yeah, I'm sure as we see more rules being converted to SwiftSyntax we'll find more common patterns and opportunities to create abstractions.